### PR TITLE
Clear monitors in case when monitoring process dies.

### PR DESCRIPTION
**Pull request by qnikst**
Assume situation when process A monitors process B. In this case
infomation about this is stored in node controller (NC).
Now when process A dies, we need to remove this information from NC
state. Otherwise we may have a memory leak when many short lived
processes monitors long lived process.

In this commit this logic is implemented - when process dies NC
traverses information about monitors and removes monitors allocated
by the process that died.

### CODE REVIEW

**Comment by qnikst** [comment]
Can't reproduce CI failures locally.

**Comment by facundominguez** [comment]
Rebased master.

**Comment by facundominguez** [comment]
I can reproduce with
`cpulimit -l 15 stack build distributed-process-tests:TestClosure`

**Comment by dcoutts** [comment]
This looks ok, but it highlights that the ncEffectDied is O(n) in the number of processes because it has to do full traversals of the links, monitors and registry mappings. It may be worth considering maintaining a bidirectional mapping, e.g. a Map ProcessId ProcInfo that contains info on what would need to be deleted for each process, so then when a process dies we can just do O(log n)-cost deletes.

**Comment by facundominguez** [comment]
@qnikst, I'm finding the bidirectional logic hard to follow. How about we organized it like this instead: https://github.com/haskell-distributed/distributed-process/compare/fd/bimultimap

**Comment by facundominguez** [comment]
After discussing it with @qnikst , we are gonna merge fd/bimultimap instead.
